From 5014728695639f33895feff9cabad199b0612c45 Mon Sep 17 00:00:00 2001
From: Paul Eggleton <paul.eggleton@microsoft.com>
Date: Sun, 8 Feb 2026 21:34:54 -0800
Subject: [PATCH] build.c: ignore return of 1 from tar -cf

When running do_package_write_deb, we have trees of hardlinked files
such as the dbg source files in ${PN}-dbg. If something makes another
copy of one of those files (or deletes one), the number of links a file
has changes and tar can notice this, e.g.:

| DEBUG: Executing python function do_package_deb
| dpkg-deb: building package `sed-ptest' in `/media/build1/poky/build/tmp/work/i586-poky-linux/sed/4.2.2-r0/deploy-debs/i586/sed-ptest_4.2.2-r0.3_i386.deb'.
| tar: ./usr/lib/sed/ptest/testsuite/tst-regex2: file changed as we read it
| dpkg-deb: error: subprocess tar -cf returned error exit status 1

Tar returns an error of 1 when files 'change' and other errors codes
in other error cases. We tweak dpkg-deb here so that it ignores an exit
code of 1 from tar. The files don't really change (and we have locking in
place to avoid that kind of issue).

Upstream-Status: Inappropriate [OE specific]

Original patch by RP 2015/3/27, rebased by Paul Eggleton

Signed-off-by: Paul Eggleton <paul.eggleton@microsoft.com>

Rebase to 1.23.5
Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 src/deb/build.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/deb/build.c b/src/deb/build.c
index f7a0156f1..67e758185 100644
--- a/src/deb/build.c
+++ b/src/deb/build.c
@@ -526,6 +526,7 @@ tarball_pack(const char *dir, filenames_feed_func *tar_filenames_feeder,
 {
 	int pipe_filenames[2], pipe_tarball[2];
 	pid_t pid_tar, pid_comp;
+	int rc;
 
 	/* Fork off a tar. We will feed it a list of filenames on stdin
 	 * later. */
@@ -591,7 +592,9 @@ tarball_pack(const char *dir, filenames_feed_func *tar_filenames_feeder,
 	 * job. */
 	close(pipe_filenames[1]);
 	subproc_reap(pid_comp, _("<compress> from tar -cf"), 0);
-	subproc_reap(pid_tar, "tar -cf", 0);
+	rc = subproc_reap(pid_tar, "tar -cf", SUBPROC_RETERROR);
+	if (rc && rc != 1)
+		ohshite(_("subprocess %s returned error exit status %d"), "tar -cf", rc);
 }
 
 static intmax_t
-- 
2.49.0

